### PR TITLE
Feat: Rollout API

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes

part of https://github.com/kubernetes-sigs/lws/issues/3

#### Does this PR introduce a user-facing change?

```release-note
Support Rollout in leaderWorkerSet.
```

### CODE REVIEW

**Comment by kerthcet** [comment]
This is API part, you can take a look if you want. cc @ahg-g @liurupeng

**Comment by ahg-g** [comment]
/lgtm

Please wait until @liurupeng reviews as well.

File: Dockerfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ COPY api/ api/
 COPY pkg/controller/ pkg/controller/
 COPY pkg/cert/ pkg/cert/
 COPY pkg/webhook/ pkg/webhook/
-COPY pkg/commonutils pkg/commonutils
```
Comment (liurupeng) [inline Dockerfile:20]: may need to do a rebase

File: api/leaderworkerset/v1/leaderworkerset_types.go [hunk 2]
```diff
@@ -89,12 +97,71 @@ type LeaderWorkerTemplate struct {
...
+	// +kubebuilder:default=1
+	MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
+
+	// The maximum number of replicas that can be scheduled above the original number of
+	// replicas.
+	// Value can be an absolute number (ex: 5) or a percentage of total replicas at
+	// the start of the update (ex: 10%).
+	// This can not be 0 if MaxUnavailable is 0.
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a value of 1 is used.
```
Comment (liurupeng) [inline api/leaderworkerset/v1/leaderworkerset_types.go:145]: if we didn't specify MaxUnavailable and MaxSurge, the default value for them are both 1? which means that actually we could have 2 replicas upgrading at the same time?

Comment (liurupeng) [reply api/leaderworkerset/v1/leaderworkerset_types.go:145]: But I think it is reasonable to have both as 1 at the same time

**Comment by ahg-g** [comment]
Before we unhold, lets discuss if we want to include this in 0.1.0 release. If not, then we should wait until we release.

**Comment by kerthcet** [comment]
Talked offline as we will release 0.1.0 first.

File: api/leaderworkerset/v1/leaderworkerset_types.go [hunk 3]
```diff
@@ -89,12 +97,71 @@ type LeaderWorkerTemplate struct {
...
+	// By default, a fixed value of 1 is used.
+	// Example: when this is set to 30%, the old replicas can be scaled down by 30%
+	// immediately when the rolling update starts. Once new replicas are ready, old replicas
+	// can be scaled down further, followed by scaling up the new replicas, ensuring
+	// that at least 70% of original number of replicas are available at all times
+	// during the update.
+	//
+	// +kubebuilder:validation:XIntOrString
+	// +kubebuilder:default=1
+	MaxUnavailable intstr.IntOrString `json:"maxUnavailable,omitempty"`
```
[the remark this answers is not in the document]
Comment (liurupeng) [reply api/leaderworkerset/v1/leaderworkerset_types.go:136]: yep, this is good

**Comment by ahg-g** [comment]
Is there an implementation PR in the works? if so, I guess we can merge this.

**Comment by kerthcet** [comment]
Yes, the implementation for `maxUnavailable & rollout` has already implemented, I'm still adding more testcases and test locally. 

For maxSurge, I'll leave for next step to reduce the review burden. We can merge now. 

/hold cancel

**Pull request merged**